### PR TITLE
Setup package

### PR DESCRIPTION
**Pull request by paolorechia**
Supports installing `exllama` as a package.

Example usage:
```
 pip install 'exllama_lib @ git+https://github.com/paolorechia/exllama@setup-package'
```

EDIT:
Worth explaining how to use the installed package. Since the installation setup creates a temporary `exllama_lib` to package `exllama`, the package also has this name, and can be imported as following:

```
from exllama_lib.model import ExLlama, ExLlamaCache, ExLlamaConfig
from exllama_lib.tokenizer import ExLlamaTokenizer
from exllama_lib.generator import ExLlamaGenerator
```

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+setup(
+    name="exllama_lib",
+    version="0.1.0",
+    author="turboderp",
+    author_email="<not provided>",
+    maintainer="turboderp",
+    maintainer_email="<not provided>",
+    license="MIT",
+    url="https://github.com/paolorechia/exllama",
```
Comment (nikshepsvn) [inline setup.py:74]: you probably want to change this too

**Comment by paolorechia** [comment]
Thanks, @nikshepsvn nikshepsvn! I fixed it, and also squashed the commits in this branch since they were quite dirty.

File: requirements.txt [hunk 2]
```diff
@@ -1,4 +1,6 @@
+-f https://download.pytorch.org/whl/nightly/cu118
 torch>=2.0.1
 safetensors==0.3.1
 sentencepiece>=0.1.97
 ninja==1.11.1
+torch>=2.0.1
```
Comment (nikuya3) [inline requirements.txt:6]: This is redundant, torch is already defined

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    name="exllama_lib",
+    version="0.1.0",
+    author="turboderp",
+    author_email="<not provided>",
+    maintainer="turboderp",
+    maintainer_email="<not provided>",
+    license="MIT",
+    url="https://github.com/turboderp/exllama",
+    description="Memory efficient implementation of llama inference.",
+    python_requires=">=3.8",
```
Comment (nikuya3) [inline setup.py:76]: Requires python >=3.9 (because of https://github.com/turboderp/exllama/issues/57#issuecomment-1595666211)

File: .gitignore
Comment (nikuya3) [inline .gitignore:1]: maybe it makes sense to unify this with `.dockerignore`

File: requirements.txt (hunk 2 above)
Comment (paolorechia) [reply requirements.txt:5]: Thanks, removed it.

File: setup.py (hunk 3 above)
Comment (paolorechia) [reply setup.py]: Thanks, updated.

File: .gitignore
Comment (paolorechia) [reply .gitignore:1]: Hi, any ideas how to do this? I've updated both files to be equal, please take a look and let me know what you think.

**Comment by paolorechia** [comment]
@nopperl thanks for the review, I've updated the PR! Sorry for the delay, I wasn't working much in my free time lately.

**Comment by Dessix** [comment]
@turboderp There's now [a reproducible Nix derivation that produces an installation of EXLlama](/TheseusInstitute/nix-exllama), which includes a `setup.py` and compiles the CUDA extension on installation instead of at runtime- with or without a GPU on the system at the time of installation. Perhaps this will be of interest?

File: cuda_ext.py [hunk 4]
```diff
@@ -43,19 +47,19 @@ def find_msvc():
...
-        os.path.join(library_dir, "exllama_ext/cuda_func/q4_matrix.cu"),
-        os.path.join(library_dir, "exllama_ext/cuda_func/q4_matmul.cu"),
-        os.path.join(library_dir, "exllama_ext/cuda_func/column_remap.cu"),
-        os.path.join(library_dir, "exllama_ext/cuda_func/rms_norm.cu"),
-        os.path.join(library_dir, "exllama_ext/cuda_func/rope.cu"),
-        os.path.join(library_dir, "exllama_ext/cuda_func/half_matmul.cu"),
-        os.path.join(library_dir, "exllama_ext/cuda_func/q4_attn.cu"),
-        os.path.join(library_dir, "exllama_ext/cuda_func/q4_mlp.cu"),
-        os.path.join(library_dir, "exllama_ext/cpu_func/rep_penalty.cpp")
+        os.path.join(library_dir, os.path.join(library_prefix, extension_name, "exllama_ext.cpp")),
```
Comment (Dessix) [inline cuda_ext.py:50]: Consider using `pathlib.Path` for these, letting you do:
```python
library_dir = Path("<...>")
extension_dir = library_dir / extension_name
[extension_dir / subpath for subpath in [
  "extllama_ext.cpp",
  "cuda_buffers.cu",
  "cuda_func/q4_matrix.cu",
  ...
]]
```
Reducing a lot of the excess line-noise wasted here.

File: requirements.txt [hunk 5]
```diff
@@ -1,4 +1,5 @@
-torch>=2.0.1
+-f https://download.pytorch.org/whl/nightly/cu118
```
Comment (Dessix) [inline requirements.txt:1]: Pinning this to Cuda 118 is unnecessary and detrimental to general usage with 12.2 having just released with C++20 support. I recommend finding an alternative and more dynamic method.

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+package_target_dir="exllama_lib"
+extension_dir = "exllama_ext"
+
+# Clean state before trying to package again
+try:
+    shutil.rmtree(package_target_dir)
+except FileNotFoundError:
+    pass
+
+os.makedirs(package_target_dir)
```
Comment (Dessix) [inline setup.py:35]: [`pathlib.Path.mkdir`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir)
```python
from pathlib import Path
Path(package_target_dir).mkdir(parents=True, exist_ok=True)
```

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        "perplexity",
+        "tokenizer"
+]
+
+for exported_file in exported_modules:
+    python_file = f"{exported_file}.py"
+    shutil.copyfile(python_file, os.path.join(package_target_dir, python_file))
+
+shutil.copytree(extension_dir, os.path.join(package_target_dir, extension_dir))
+pathlib.Path(os.path.join(package_target_dir, "__init__.py")).touch(exist_ok=True)
```
Comment (Dessix) [inline setup.py:52]:
```suggestion
(pathlib.Path(package_target_dir) / "__init__.py").touch(exist_ok=True)
```

**Review by Dessix — CHANGES_REQUESTED** [review]
I recommend switching most usages of `os.path` to `pathlib`, which is more flexible cross-platform and more concise in general.

**Comment by paolorechia** [comment]
Sorry, @Dessix, but I’m not planning to apply the requested changes. I appreciate the review and understand the raised issues but they go beyond the original intent of this PR.

Feel free to propose a different PR to replace this one.

**Comment by paolorechia** [comment]
Closing since it's stale.

**Pull request closed without merging**